framework/decode/vulkan_replay_resource_dump_json.h

```diff
@@ -0,0 +1,119 @@
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef GFXRECON_VULKAN_REPLAY_DUMP_RESOURCE_DUMP_H
+#define GFXRECON_VULKAN_REPLAY_DUMP_RESOURCE_DUMP_H
+
+#include "decode/json_writer.h"
+#include "project_version.h"
+#include "util/file_output_stream.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+class VulkanReplayResourceDumpJson {
+public:
+  VulkanReplayResourceDumpJson(){};
+
+  void VulkanReplayResourceDumpJsonOpen(const std::string &infile) {
+    std::string outfile;
+    int i = infile.size();
+    int i2 = infile.compare(infile.size() - 4, 5, ".gfxr");
+    if (infile.size() >= 5 && infile.compare(infile.size() - 4, 5, ".gfxr")) {
+      outfile = infile.substr(0, infile.size() - 5);
+    }
+    outfile = outfile + "_rd.json";
+
+    gfxrecon::util::platform::FileOpen(&jsonFileHandle_, outfile.c_str(), "w");
+    assert(jsonFileHandle_); // TODO: Generate an error if FileOpen fails
+    out_stream_ =
+        new gfxrecon::util::FileNoLockOutputStream(jsonFileHandle_, false);
+    gfxrecon::util::JsonOptions json_options;
+    json_options.root_dir = ".";     // This field is not used
+    json_options.data_sub_dir = "."; // This field is not used
+    json_options.format = gfxrecon::util::JsonFormat::JSON;
+
+    json_writer_ = new gfxrecon::decode::JsonWriter(
+        json_options, GFXRECON_PROJECT_VERSION_STRING, infile);
+
+    json_writer_->GetHeaderJson()["vulkan-version"] =
+        std::to_string(VK_VERSION_MAJOR(VK_HEADER_VERSION_COMPLETE)) + "." +
+        std::to_string(VK_VERSION_MINOR(VK_HEADER_VERSION_COMPLETE)) + "." +
+        std::to_string(VK_VERSION_PATCH(VK_HEADER_VERSION_COMPLETE));
+    json_writer_->StartStream(out_stream_);
+    dump_ = new nlohmann::ordered_json();
+  }
+
+  void VulkanReplayResourceDumpJsonBlockStart() {
+    json_data_ = &json_writer_->WriteBlockStart();
+  }
+
+  void VulkanReplayResourceDumpJsonBlockEnd() {
+    static uint32_t n = 0;
+    (*json_data_)["resourceDump_" + std::to_string(n++)] = *dump_;
+    json_writer_->WriteBlockEnd();
+  }
+
+  void VulkanReplayResourceDumpJsonData(std::string descriptor,
+                                        std::string value) {
+    (*dump_)[descriptor] = value;
+  }
+
+  void VulkanReplayResourceDumpJsonData(std::string descriptor,
+                                        uint64_t value) {
+    (*dump_)[descriptor] = value;
+  }
+
+  // TODO:
+  // Should move this code to the destructor and delete this method. But it's a
+  // a separate method because VulkanReplayResourceDump is calling exit(0),
+  // which results in destructors not being called and the json file not getting
+  // closed out. VulkanReplayResourceDump explicity calls this method before
+  // calling exit(0).
+  void VulkanReplayResourceDumpJsonClose() {
+    if (json_writer_) {
+      json_writer_->EndStream();
+      gfxrecon::util::platform::FileClose(jsonFileHandle_);
+
+      //@@@ TODO: Not sure if this is the best way to release these. Could I
```

Comment:
Do these variables need to be pointers? If so perhaps they can be smart pointers (unique_ptr) so they will be deleted by themselves once the destructor is called?